codalab/worker/download_util.py

```diff
@@ -125,9 +155,100 @@ def _compute_target_info(path, depth):
         result['type'] = 'directory'
         if depth > 0:
             result['contents'] = [
-                _compute_target_info(os.path.join(path, file_name), depth - 1)
+                _compute_target_info_local(os.path.join(path, file_name), depth - 1)
                 for file_name in os.listdir(path)
             ]
     if result is None:
         raise PathException()
     return result
+
+
+def _compute_target_info_beam(path, depth):
+    """Computes target info for a file that is externalized on a location
+    such as Azure, by using the Apache Beam FileSystem APIs."""
+
+    linked_bundle_path = parse_linked_bundle_url(path)
+    if not FileSystems.exists(linked_bundle_path.bundle_path):
+        raise PathException(linked_bundle_path.bundle_path)
+    if not linked_bundle_path.is_archive:
+        # Single file
+        file = FileSystems.match([path])[0].metadata_list[0]
+        return {
+            'name': linked_bundle_path.bundle_uuid,
+            'type': 'file',
+            'size': file.size_in_bytes,
+            'perm': 0o755,
+        }
+
+    tf, _ = open_indexed_tar_gz_file(linked_bundle_path.bundle_path)
+    islink = lambda finfo: stat.S_ISLNK(finfo.mode)
+    readlink = lambda finfo: finfo.linkname
+
+    isfile = lambda finfo: finfo.type in tarfile.REGULAR_TYPES
+    isdir = lambda finfo: finfo.type == tarfile.DIRTYPE
+    listdir = lambda path: tf.getFileInfo(path, listDir=True)
+
+    def _get_info(path, depth):
+        if not path.startswith("/"):
+            path = "/" + path
+        finfo = tf.getFileInfo(path)
+        if finfo is None:
+            # Not found
+            raise PathException
+        result = {}
+        result['name'] = path.split("/")[-1]  # get last part of path
+        result['size'] = finfo.size
+        result['perm'] = finfo.mode & 0o777
+        if islink(finfo):
+            result['type'] = 'link'
+            result['link'] = readlink(finfo)
+        elif isfile(finfo):
+            result['type'] = 'file'
+        elif isdir(finfo):
+            result['type'] = 'directory'
+            if depth > 0:
+                result['contents'] = [
+                    _get_info(path + "/" + file_name, depth - 1)
+                    for file_name in listdir(path)
+                    if file_name != "."
+                ]
+        return result
+
+    if linked_bundle_path.archive_subpath:
+        # Return the contents of a subpath within a directory.
+        return _get_info(linked_bundle_path.archive_subpath, depth)
+    else:
+        # No subpath, return the entire directory.
+        file = FileSystems.match([path])[0].metadata_list[0]
+        result = {
+            'name': linked_bundle_path.bundle_uuid,
+            'type': 'directory',
+            'size': file.size_in_bytes,
+            'perm': 0o755,
```

Comment:
Same as above...permissions should come from bundle?

Reply:
This code path is only called when someone gets the target info of the _entire bundle_. The _entire bundle_ technically doesn't have a file permission -- only the constituent files within the bundle do.

Reply:
So 0o755 doesn't actually make a difference or is used by the code -- it's just a placeholder in some sense.

Reply:
Looking at the analogous situation for disk-based storage -- the permission of the bundle directory isn't configurable by the user, either. It's just a constant value determined by the umask setting of the system. That's why hardcoding the permission to 0o755 here would accomplish equivalent behavior.